lib/netutil/proxyprotocol.go

```diff
@@ -0,0 +1,176 @@
+package netutil
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+)
+
+var bufioReaderPool = sync.Pool{}
+
+func getBufioReader(r io.Reader) *bufio.Reader {
+	v := bufioReaderPool.Get()
+	if v == nil {
+		return bufio.NewReader(r)
+	}
+	br := v.(*bufio.Reader)
+	br.Reset(r)
+	return br
+}
+
+func putBufioReader(r *bufio.Reader) {
+	bufioReaderPool.Put(r)
+}
+
+type proxyProtocolConn struct {
+	net.Conn
+	br         *bufio.Reader
+	remoteAddr net.Addr
+}
+
+func newProxyProtocolConn(srcConn net.Conn) (net.Conn, error) {
+	br := getBufioReader(srcConn)
+	maybeSrcAddr, err := readProxyProto(br)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read proxy protocol from connection: %w", err)
+	}
+	if maybeSrcAddr == nil {
+		maybeSrcAddr = srcConn.RemoteAddr()
+	}
+	return &proxyProtocolConn{
+		Conn:       srcConn,
+		br:         br,
+		remoteAddr: maybeSrcAddr,
+	}, nil
+}
+
+func (ppc *proxyProtocolConn) Read(b []byte) (n int, err error) {
+	return ppc.br.Read(b)
+}
+
+func (ppc *proxyProtocolConn) RemoteAddr() net.Addr {
+	return ppc.remoteAddr
+}
+
+func (ppc *proxyProtocolConn) Close() error {
+	putBufioReader(ppc.br)
+	return ppc.Conn.Close()
+}
+
+var (
+	V2Identifier = []byte("\r\n\r\n\x00\r\nQUIT\n")
+)
+
+// https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt
+func readProxyProto(r *bufio.Reader) (net.Addr, error) {
+	// first 12 bytes - signature
+	// 13 byte - protocol and command
+	maybeProto, err := r.Peek(13)
```

Comment:
So the first 12 bytes were already read before? If yes, why mention it in a comment?

Reply:
The initial idea was to automatically determine whether the proxy protocol is used or not for the newly established connection. Unfortunately, this doesn't work in the case if the connection doesn't use proxy protocol and the client sends less than 13 bytes to the connection. In this case the server will hang indefinitely at the the line 73. @f41gh7 , fyi. So the updated version of this pull request accepts either plain connections or proxy connections depending on the initial settings passed to the constructor of the TCP listener.

The updated version also removes buffering for the connection, since the accepted connection is wrapped into buffer later. The updated version makes sure that the proxy protocol is read with 2 recv() syscalls - the first syscall reads the first 16 bytes (aka header) and the second syscall reads the remaining payload with the length encoded in the last 2 bytes of the header.